### ISSUE TITLE
Properties from mapped type have no declarations.

### ISSUE THREAD

**Issue by aam051102** [dsherret/ts-simple-ast#1315]
**Describe the bug**

Version: 15.1.0

When trying to figure out the resulting type from a mapped type, attempting to get the declarations/types of the properties in the mapped type, will only result in an empty array. Note that this only seems to occur when using an "as" constraint for the key.

**To Reproduce**

Executing the following code will log an empty declaration array for the "a" property in type "B", as mapped by "y" in type "A".

```ts
const { Project, Node, TypeNode } = require("ts-morph");

const project = new Project();
const sourceFile = project.createSourceFile(
    "test.ts",
    `
type B = {
	a: string;
	c: boolean;
};

type A = {
    y: {
        [Key in keyof B as Key extends "c" ? never : Key]: B[Key] | null;
    }
}

const x: A;
`
);

const varDeclaration = sourceFile.getVariableDeclarationOrThrow("x");
const varType = varDeclaration.getType();

if (varType.isObject()) {
    const declaration = varType.getProperties()[0].getDeclarations()[0];

    if (Node.isPropertySignature(declaration)) {
        const tn = declaration.getTypeNode();
        
        if (TypeNode.isMappedTypeNode(tn)) {
            const t = tn.getType();

            if (t.isObject()) {
                const aType = t.getProperties()[0];

                console.log(aType.getDeclarations()); // Returns []
            }
        }
    }
}
```

**Expected behavior**

I expect to get either a new, merged type or, at the very least, a union type with null and some sort of structure with the outputted B[Key] types. If they can be found elsewhere already, that's fine, but I can't seem to find them anywhere.

**Comment by Gerrit0** [comment]
I'm pretty sure this is how the compiler API works, so ts-morph is just mirroring that behavior... you can still work with the type side of things, but requesting declarations (e.g. user source code location for the property) won't give anything since there isn't one.

**Comment by aam051102** [comment]
@Gerrit0 I can understand that there wouldn't be any declarations if that's the case, but how would I then get the types? I tried to use aType.getDeclaredType(), since that seems to be the only type getter I have available, but it returns an error object. Is there another way to get the correct type?

**Comment by Gerrit0** [comment]
```ts
console.log(aType.getTypeAtLocation(declaration).getText());
```

Should do it

**Comment by aam051102** [comment]
@Gerrit0 Well, it does get me a type, but only the string part. It's not a union type of "string | null", like I would expect. I can see that it gets me the right type if I define it without the indexing in the mapped type (e.x. "null" instead of "B[Key] | null" returns "null"). Why is there a difference between the two? Isn't there a way to simply get the completed type?

EDIT
A clearer example:
`[Key in keyof B as Key extends "c" ? never : Key]: B[Key] | null;` - expected "string | null", but got "string"
`[Key in keyof B as Key extends "c" ? never : Key]: null;` - expected "null" and got "null"

EDIT 2
On further investigation, it seems to get the right type when using values other than null in the union type? Is the null from the union hidden in a function somewhere if it's not in the type? Where do I find it?

In other words, this works:
`[Key in keyof B as Key extends "c" ? never : Key]: B[Key] | number;` - expected "string | number" and got "string | number"
But this doesn't:
`[Key in keyof B as Key extends "c" ? never : Key]: B[Key] | null;` - expected "string | null", but got "string"

**Comment by aam051102** [comment]
@Gerrit0 As a follow-up to my previous message, I have now found the reason for "null" missing in [this issue](https://github.com/dsherret/ts-morph/issues/652). `getTypeAtLocation()` seems to be the right solution. Thanks for the help.
